cpp/src/arrow/type_traits.h

```diff
@@ -22,7 +22,9 @@
 #include <type_traits>
 #include <vector>
 
+#include "arrow/array/data.h"
```

Comment:
Hmm, it would be better if we didn't add a new inclusion here, as this header is bound to be frequently included.

Reply:
This is a sign some of the types being used here haven't been forward-declared in `type_fwd.h`